dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java

```diff
@@ -733,4 +721,61 @@ protected TrackedEntityInstance postProcessObject( TrackedEntityInstance tracked
     {
         return (trackedEntityInstance == null || trackedEntityInstance.isDeleted()) ? null : trackedEntityInstance;
     }
+
+    @Override
+    public Optional<TrackedEntityInstance> getTrackedEntityInstanceWithUniqueAttributeValue( TrackedEntityInstanceQueryParams params )
+    {
+        // ---------------------------------------------------------------------
+        // Select clause
+        // ---------------------------------------------------------------------
+
+        String sql = "select tei.uid as uid from trackedentityinstance tei ";
+
+        if ( params.hasOrganisationUnits() )
+        {
+            sql += "inner join organisationunit ou on tei.organisationunitid = ou.organisationunitid " +
+                "and ou.uid in (" + params.getOrganisationUnits() + ") ";
+        }
+
+        for ( QueryItem item : params.getAttributes() )
+        {
+            final String col = statementBuilder.columnQuote( item.getItemId() );
+
+            sql += "inner join trackedentityattributevalue as " + col + " on " + col
+                + ".trackedentityinstanceid = tei.trackedentityinstanceid and " + col
+                + ".trackedentityattributeid = " + item.getItem().getId() + " ";
+
+            for ( QueryFilter filter : item.getFilters() )
+            {
+                final String encodedFilter = statementBuilder.encode( filter.getFilter(), false );
+                final String queryCol = item.isNumeric() ? (col + ".value") : "lower(" + col + ".value)";
+
+                sql += "and " + queryCol + " LIKE "
+                    + StringUtils.lowerCase( filter.getSqlFilter( encodedFilter ) ) + " ";
+            }
+        }
+
+        if ( !params.isIncludeDeleted() )
+        {
+            sql += "where tei.deleted is false ";
+        }
+
+        sql += " limit 1";
+
+
+        // ---------------------------------------------------------------------
+        // Query
+        // ---------------------------------------------------------------------
+        SqlRowSet rowSet = jdbcTemplate.queryForRowSet( sql );

```

Comment:
I'm not sure it's a good idea to use `jdbcTemplate` here, since you risk to query data which has not been flushed yet by Hibernate. You should consider using:
1) `session.createSQLQuery(..)` -> this will force Hibernate to flush
2) convert your query into HQL

I think (2) is the best approach, since we may have a performance penalty for flush.

Reply:
Fixed.